File: PseudoCoupHQ/Research/oracle/riscv/softfloat_slices/emulations/rust/f16_roundToInt.rs
#![allow(non_snake_case, unused_parens, unused_imports, unused_variables, unused_comparisons, clippy::all)]
use crate::helpers::*;


pub fn f16_roundToInt_rm1(v_arg: u64, v_arg1: bool) -> u64 {
    let v_i: u16 = ((v_arg) as u16);
    let v_i2: u32 = ((v_arg) as u32);
    let v_i3: u32 = v_i2.wrapping_shr((0xau32) as u32);
    let v_i4: u32 = (v_i3 & 0x1fu32);
    let v_i14: u32 = (v_i2 & 0x3ffu32);
    let v_i15: bool = (v_i14 == 0x0u32);
    let v_i5: bool = (v_i4 < 0xfu32);
    let v_i11: bool = (v_i4 > 0x18u32);
    let v_i7: u64 = (v_arg & 0x7fffu64);
    let v_i8: bool = (v_i7 == 0x0u64);
    let v__m1: u16 = (0u16.wrapping_sub((v_i8) as u16));
    let v_i9: u16 = (v_i & 0x8000u16);
    let v__a2: u16 = (v_i & v__m1);
    let v__n3: u16 = (v__m1 ^ 0xffffu16);
    let v__a4: u16 = (v_i9 & v__n3);
    let v_spec_select: u16 = (v__a2 | v__a4);
    let v_i13: bool = (v_i4 != 0x1fu32);
    let v_i16: bool = (v_i15 | v_i13);
    let v__m5: u16 = (0u16.wrapping_sub((v_i16) as u16));
    let v_i18: u32 = 0x19u32.wrapping_sub(v_i4);
    let v__sh9: u32 = 0xffffu32.wrapping_shl((v_i18) as u32);
    let v_i19: u32 = v__sh9;
    let v_i20: u16 = ((v_i19) as u16);
    let v_i21: u16 = (v_i & v_i20);
    let v__a6: u16 = (v_i & v__m5);
    let v__n7: u16 = (v__m5 ^ 0xffffu16);
    let v__a8: u16 = (0x7e00u16 & v__n7);
    let v_spec_select4: u16 = (v__a6 | v__a8);
    let v__n10: bool = (v_i11 ^ true);
    let v__n11: bool = (v_i5 ^ true);
    let v__c12: bool = (v__n10 & v__n11);
    let v__m13: u16 = (0u16.wrapping_sub((v__c12) as u16));
    let v__a14: u16 = (v_i21 & v__m13);
    let v__m15: u16 = (0u16.wrapping_sub((v_i5) as u16));
    let v__a16: u16 = (v_spec_select & v__m15);
    let v__o17: u16 = (v__a14 | v__a16);
    let v__m19: u16 = (0u16.wrapping_sub((v_i11) as u16));
    let v__a20: u16 = (v_spec_select4 & v__m19);
    let v__o21: u16 = (v__o17 | v__a20);
    let v_i23: u64 = ((v__o21) as u64);
    v_i23
}

File: PseudoCoupHQ/Research/oracle/riscv/softfloat_slices/emulations/js/f16_roundToInt.js
'use strict';
const { sfUdiv, sfSgn, sfCtlz, sfAbs, sfUsubsat, sfFshl } = require('./helpers.js');


function f16_roundToInt_rm1(v_arg, v_arg1) {
  const v_i = (v_arg & 0xffffn);
  const v_i2 = (v_arg & 0xffffffffn);
  const v_i3 = (v_i2 >> (0xan & 0x1fn));
  const v_i4 = (v_i3 & 0x1fn);
  const v_i14 = (v_i2 & 0x3ffn);
  const v_i15 = ((v_i14 === 0x0n) ? 1n : 0n);
  const v_i5 = ((v_i4 < 0xfn) ? 1n : 0n);
  const v_i11 = ((v_i4 > 0x18n) ? 1n : 0n);
  const v_i7 = (v_arg & 0x7fffn);
  const v_i8 = ((v_i7 === 0x0n) ? 1n : 0n);
  const v__m1 = (sfSgn(v_i8, 0x1n) & 0xffffn);
  const v_i9 = (v_i & 0x8000n);
  const v__a2 = (v_i & v__m1);
  const v__n3 = (v__m1 ^ 0xffffn);
  const v__a4 = (v_i9 & v__n3);
  const v_spec_select = (v__a2 | v__a4);
  const v_i13 = ((v_i4 !== 0x1fn) ? 1n : 0n);
  const v_i16 = (v_i15 | v_i13);
  const v__m5 = (sfSgn(v_i16, 0x1n) & 0xffffn);
  const v_i18 = ((0x19n - v_i4) & 0xffffffffn);
  const v__sh9 = ((0xffffn << (v_i18 & 0x1fn)) & 0xffffffffn);
  const v_i19 = v__sh9;
  const v_i20 = (v_i19 & 0xffffn);
  const v_i21 = (v_i & v_i20);
  const v__a6 = (v_i & v__m5);
  const v__n7 = (v__m5 ^ 0xffffn);
  const v__a8 = (0x7e00n & v__n7);
  const v_spec_select4 = (v__a6 | v__a8);
  const v__n10 = (v_i11 ^ 0x1n);
  const v__n11 = (v_i5 ^ 0x1n);
  const v__c12 = (v__n10 & v__n11);
  const v__m13 = (sfSgn(v__c12, 0x1n) & 0xffffn);
  const v__a14 = (v_i21 & v__m13);
  const v__m15 = (sfSgn(v_i5, 0x1n) & 0xffffn);
  const v__a16 = (v_spec_select & v__m15);
  const v__o17 = (v__a14 | v__a16);
  const v__m19 = (sfSgn(v_i11, 0x1n) & 0xffffn);
  const v__a20 = (v_spec_select4 & v__m19);
  const v__o21 = (v__o17 | v__a20);
  const v_i23 = (v__o21);
  return v_i23;
}
module.exports = { f16_roundToInt_rm1 };
